docs/contract/confidential-smart-contract.md

```diff
@@ -0,0 +1,264 @@
+        let rsp = HelloWorld::call(
+            &mut ctx,
+            Request::SayHello {
+                who: "unit test".to_string(),
+            },
+        )
+        .expect("SayHello call should work");
+
+        // Make sure the greeting is correct.
+        assert_eq!(
+            rsp,
+            Response::Hello {
+                greeting: "hello unit test (11)".to_string()
+            }
+        );
+
+        // Dispatch another SayHello message.
+        let rsp = HelloWorld::call(
+            &mut ctx,
+            Request::SayHello {
+                who: "second call".to_string(),
+            },
+        )
+        .expect("SayHello call should work");
+
+        // Make sure the greeting is correct.
+        assert_eq!(
+            rsp,
+            Response::Hello {
+                greeting: "hello second call (12)".to_string()
+            }
+        );
+    }
+}
+```
+
+The contract is built the same way as its non-confidential counterpart:
+
+```shell
+cargo build --target wasm32-unknown-unknown --release
+```
+
+::: warning
+
+The blockchain store containing all compiled contracts is public. This means
+that anyone will be able to decompile your smart contract and see how it
+works. **Do not put any sensitive data inside the smart contract code!** 
+
+:::
+
+Since the smart contracts store is public, uploading the WASM code is
+the same as for the non-confidential ones:
+
+```shell
+oasis contracts upload hello_world.wasm
+```
+
+## Confidential Instantiation and Calling
+
+To generate a confidential transaction, the `oasis contracts` subcommand
+accepts `--encrypted` flag. Confidential transactions have encrypted contract
+address, function name, parameters and the amounts and types of tokens sent.
+**However, the _authorization information_ which contains information on the
+signer is public!** Usually this field contains the public key of your account
+or a list of expected multisig keys.
+
+::: danger
+
+While the transaction itself is confidential, the effects of a smart contract
+execution may reveal some information. For example, the account balances are
+public. If the effect is, say, subtraction of 10 tokens from the signer's
+account, this most probably implies that they have been transferred as part of
+this transaction.
+
+:::
+
+To instantiate a smart contract with the code ID obtained above, you can run:
+
+```shell
+oasis contracts instantiate CODEID '{instantiate: {initial_counter: 42}}' --encrypted
+```
+
+As is the case with non-confidential contracts, we obtain an incremented
+instance ID. Finally, we make a confidential call:
+
+```shell
+oasis contracts call INSTANCEID '{say_hello: {who: "me"}}' --encrypted
+```
+
+::: danger
+
+Regardless of the confidential storage used in the smart contract, any [emitted
+event][emit_event] will be public.
+
+:::
+
+[emit_event]:
+  https://github.com/oasisprotocol/oasis-sdk/blob/main/runtime-sdk/src/context.rs#L804-L808
```

Comment:
Since we are in a contract tutorial it would be better to link to the [contract-sdk context](https://api.docs.oasis.dev/oasis-sdk/oasis_contract_sdk/context/trait.Context.html#tymethod.emit_event). Also note that we can link to our published Rustdoc.

Reply:
Yeah, I also added rustdoc links to PublicCell and ConfidentialCell above.